fenicsxprecice/adapter_core.py

```diff
@@ -122,58 +155,210 @@ def convert_fenicsx_to_precice(fenicsx_function, local_coords):
+def get_fenicsx_interpolation_points(function_space: fem.FunctionSpace, coupling_boundary, comm: MPI.Comm):
     """
-    Extracts vertices which FEniCSx accesses on this rank and which lie on the given coupling domain, from a given
-    function space.
+    Determines the interpolation points FEniCSx needs to interpolate the coupling boundary and the coordinates for the preCICE mesh.
 
     Parameters
     ----------
-    function_space : FEniCSx function space
-        Function space on which the finite element problem definition lives.
-    coupling_subdomain : FEniCSx Domain
-        Subdomain consists of only the coupling interface region.
-    dims : int
-        Dimension of problem.
+    function_space:  fem.FunctionSpace
+        The function space of the problem
+    coupling_boundary:
+        A callable function describing the coupling boundary
+    comm: MPI.Comm
+        The used MPI communicator
 
     Returns
     -------
-    ids : numpy array
-        Array of ids of fenicsx vertices.
-    coords : numpy array
-        The coordinates of fenicsx vertices in a numpy array [N x D] where
-        N = number of vertices and D = dimensions of geometry.
+    (ndarray, list, dict):
+        Returns a triplet of (interpolation coordinates, interpolation cells, function values to be sent to other MPI ranks)
     """
+    comm_size = comm.Get_size()
+    comm_rank = comm.Get_rank()
+
+    # domain of function space
+    domain = function_space.mesh
+    # function used to query interpolation coordinates
+    query_function = fem.Function(function_space)
+
+    # variables and function definition required for querying the coordinates FEniCSx needs
+    interpolation_coordinates = []
+    vec_len = function_space.dofmap.bs
+
+    def query_coordinates(x):
+        # to avoid UnboundLocalError, interpolation_coordinates is an array to which x is appended to
+        interpolation_coordinates.append(np.transpose(copy.deepcopy(x)))
+        # directly round and make each rounded coordinate unique to keep the code clean and to reduce memory consumption
+        interpolation_coordinates[-1] = round_unique_coordinates(interpolation_coordinates[-1])
+
+        return np.zeros((vec_len, x.shape[1]))
+
+    # determine process local cells that are on the coupling boundary
+    dofs_coupling = fem.locate_dofs_geometrical(function_space, coupling_boundary)
+    dofs_coupling_coordinates = function_space.tabulate_dof_coordinates()[dofs_coupling]
+    bb_tree = geometry.bb_tree(domain, domain.geometry.dim)
+    cell_candidates_local = geometry.compute_collisions_points(bb_tree, dofs_coupling_coordinates)
+    cell_candidates_local = geometry.compute_colliding_cells(domain, cell_candidates_local, dofs_coupling_coordinates)
+    # the cell candidates with local cell ids
+    cell_candidates_local = np.unique(cell_candidates_local.array)
+
+    # determine process-owned cells
+    index_map = domain.topology.index_map(domain.topology.dim)
```

Comment:
Looks odd because `domain.topology` should have access to `dim` internally.

Reply:
Okay. I meant it more from the software perspective that it looks like you are passing the `dim` variable of `domain.topology` into a function, which is again of `domain.topology`. This is not too important, though, so let's leave it like it is if it works.